common/SolidityTestUtils.js

```diff
@@ -9,6 +9,98 @@ async function didContractThrow(promise) {
   return false;
 }
 
+async function minePendingTransactions(web3, time) {
+  return new Promise((resolve, reject) => {
+    web3.currentProvider.send(
+      {
+        jsonrpc: "2.0",
+        method: "evm_mine",
+        params: [time],
+        id: new Date().getTime()
+      },
+      (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result);
+        }
+      }
+    );
+  });
+}
+
+async function stopMining(web3) {
+  return new Promise((resolve, reject) => {
+    web3.currentProvider.send(
+      {
+        jsonrpc: "2.0",
+        method: "miner_stop",
+        params: [],
+        id: new Date().getTime()
+      },
+      (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result);
+        }
+      }
+    );
+  });
+}
+
+async function startMining(web3) {
+  return new Promise((resolve, reject) => {
+    web3.currentProvider.send(
+      {
+        jsonrpc: "2.0",
+        method: "miner_start",
+        params: [],
+        id: new Date().getTime()
+      },
+      (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result);
+        }
+      }
+    );
+  });
+}
+
+// This function will mine all transactions in `transactions` in the same block with block timestamp `time`.
+// The return value is an array of receipts corresponding to the transactions.
+// Each transaction in transactions should be a web3 transaction generated like:
+// let transaction = truffleContract.contract.methods.myMethodName(arg1, arg2);
+// or if already using a web3 contract object:
+// let transaction = web3Contract.methods.myMethodName(arg1, arg2);
+async function mineTransactionsAtTime(web3, transactions, time, sender) {
+  await stopMining(web3);
+
+  const receiptPromises = [];
+  for (const transaction of transactions) {
+    const result = transaction.send({ from: sender });
+
+    // Awaits the transactionHash, which signifies the transaction was sent, but not necessarily mined.
+    await new Promise((resolve, reject) => {
+      result.on("transactionHash", function() {
+        resolve();
+      });
+    });
+
+    // result, itself, is a promise that will resolve to the receipt.
+    receiptPromises.push(result);
+  }
+
+  await minePendingTransactions(web3, time);
```

Comment:
Just so I understand: Irrespective of the mining running (`miner_start`) or the mining stopped (`miner_stop`) this call will force the blocks to be mined at the given time stamp due to `evm_mine`, right? However as the promises have not yet been resolved the transactions have yet to be submitted to Ganache at this point so what exactly does this do? I am basing this off [this](https://github.com/trufflesuite/ganache-core#custom-methods).

Reply:
[This](https://github.com/trufflesuite/ganache-cli/issues/108#issuecomment-250600965) is where I got the flow that I used. If you don't stop mining before sending `evm_mine`, the transactions will be mined instantly by the autominer. If you stop mining, they'll be queued and dumped into the next block IIUC.

Reply:
Does that make sense?

Reply:
Totally. you stop it so that `evm_mine` does not auto mine. then you enque them then you mine them all at once. Perfect 👍